### PR TITLE
Utility private ShaderPass class to abstract some code scattered in the codebase

### PR DESCRIPTION
**Pull request by mvaligursky**
Related to https://github.com/playcanvas/engine/issues/4325 (does not handle the renaming part yet)

### CODE REVIEW

File: src/scene/shader-pass.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import { SHADER_SHADOW, SHADOW_COUNT, LIGHTTYPE_COUNT } from './constants.js';
+
+/**
+ * A pure static utility class, responsible for math operations on the shader pass constants.
+ *
+ * @ignore
+ */
+class ShaderPass {
+    /**
+     * Rturns true if the shader pass is a shadow shader pass.
```
Comment (GSterbrant) [inline src/scene/shader-pass.js:11]: Returns

Comment (mvaligursky) [reply src/scene/shader-pass.js:11]: thanks!

**Comment by slimbuck** [comment]
This is so much better!

Though TBH I'm now wondering why we store the precise type of shadow generation as a parse! Seems like SHADER_PARSE should be 'shadow depth' and the shadow type should be stored separately.

This should simplify all the code that does parse-specific logic.

**Comment by slimbuck** [comment]
(This is actually a question @mvaligursky - what do you think?)

**Review by GSterbrant — APPROVED** [review]

File: src/scene/shader-pass.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+     */
+    static toShadowType(pass) {
+        Debug.assert(ShaderPass.isShadow(pass));
+        const shadowMode = pass - SHADER_SHADOW;
+        const lightType = Math.floor(shadowMode / SHADOW_COUNT);
+        return shadowMode - lightType * SHADOW_COUNT;
+    }
+
+    /**
+     * Returns a shader pass for specified light and shadow types.
```
Comment (willeastcott) [inline src/scene/shader-pass.js:46]:
```suggestion
     * Returns a shader pass for specified light and shadow types.
```

**Review by slimbuck — APPROVED** [review]

**Pull request merged**